projects/hathor-wallet-headless/hsm-c-bridge.md

```diff
@@ -0,0 +1,345 @@
+| D_RECV_FAILED            |     -14      |   NETWORK_ERROR    |
+| D_SETSOCKOPT_FAILED      |     -16      |   NETWORK_ERROR    |
+| D_CRL_OPERATION_TIMEDOUT |     105      |   NETWORK_ERROR    |
+| D_CRL_SEND_ERROR         |     108      |   NETWORK_ERROR    |
+| D_CRL_RECV_ERROR         |     109      |   NETWORK_ERROR    |
+| ERR_NET_FAIL             |     5001     |   NETWORK_ERROR    |
+
+All other errors should be passed to the Node.js client as fatal errors, along with their Dinamo return codes for easier debugging.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+### Implementation by third-party
+This Bridge approach will be implemented by the Hathor Labs team, not the Dinamo Networks one that is the actual code owner. That way, the possibility of having minor bugs or documentation misunderstandings exist.
+
+The Dinamo Networks team is also working on a Javascript implementation that could be imported as a library through _npm_ directly to the Headless Wallet application, avoiding many of the issues of this simpler approach proposed here. However, the ETA for this library is still undefined.
+
+### Multiple processes running on docker
+For applications running on Docker environments, such as the Hathor Headless Wallet, it would be necessary to invoke the C executables from the NodeJS main process. This is not a recommended best practice as mentioned on the [_Decouple Applications_ guideline](https://docs.docker.com/develop/develop-images/guidelines/#decouple-applications).
+
+This could be solved by compiling the C files as _shared object_ (`.so`) file and invoke it directly from the NodeJS through dependencies such as [`node-addon-api`](https://www.npmjs.com/package/node-addon-api) or [`node-ffi-napi`](https://www.npmjs.com/package/ffi-napi). ( Reference [article](https://medium.com/ai-innovation/a-guide-for-javascript-developers-to-build-c-add-ons-with-node-addon-api-28c84a0c0cb1)) ).
+
+By using this approach, it would only be necessary, in each Docker environment with access to the HSM:
+- Install the Dinamo libraries on the instance
+- Copy the C files provided by Hathor on a temporary folder
+- Compile them into shared objects
+- Move the compiled objects to a folder within the headless application one
+- Remove the temporary C files folder
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+### Bridge communication through TCP
+A more standardized way of communicating with the javascript applications would be through common HTTP and websocket channels.
+
+This would mean this bridge would need to become a standalone daemon and webserver application instead of a single request process, requiring a dockerized environment for itself.
+
+That would increase its maintainability at the cost of the added complexity both for code construction and process maintenance.
+
+Because of that, this solution was discarded as the first version of this client.
+
+### NodeJs to Shared Object communication
```

Comment:
SInce this is just an alternatives sections maybe we can include that the .so source doesn't need to be in C.
We could build the bridge in rust while calling the C lib from dynamo and exposing just the rust lib.
From here we could build for any language, for instance javascript with napi-rs or python with pyo3